imagepuller/internal/service/service.go

```diff
@@ -0,0 +1,142 @@
+// Copyright 2025 Edgeless Systems GmbH
+// SPDX-License-Identifier: BUSL-1.1
+
+package service
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"log/slog"
+	"os"
+	"path/filepath"
+
+	"github.com/containers/storage"
+	"github.com/containers/storage/types"
+	"github.com/edgelesssys/contrast/imagepuller/internal/api"
+	"github.com/google/go-containerregistry/pkg/name"
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+	"github.com/google/go-containerregistry/pkg/v1/remote/transport"
+	"golang.org/x/sys/unix"
+)
+
+// ImagePullerService is the struct for which the PullImage ttRPC service is implemented.
+type ImagePullerService struct {
+	Logger *slog.Logger
+}
+
+// PullImage is a ttRPC service which pulls and mounts docker images.
+func (s *ImagePullerService) PullImage(ctx context.Context, r *api.ImagePullRequest) (response *api.ImagePullResponse, retErr error) {
+	log := s.Logger.With(
+		slog.String("image_url", r.ImageUrl),
+		slog.String("bundle_path", r.BundlePath),
+	)
+	log.Info("Handling image pull request")
+
+	defer func() {
+		if retErr != nil {
+			log.Error("request failed", "err", retErr)
+			response = nil
+		}
+	}()
+
+	ref, err := name.ParseReference(r.ImageUrl)
+	if err != nil {
+		return nil, fmt.Errorf("parsing the image URL as a references: %w", err)
+	}
+
+	tr := transport.NewRetry(remote.DefaultTransport)
+	remoteImg, err := remote.Image(ref, remote.WithContext(ctx), remote.WithTransport(tr))
+	if errors.Is(err, context.Canceled) || errors.Is(ctx.Err(), context.Canceled) {
+		return nil, fmt.Errorf("pull aborted (context cancelled): %w, %w", ctx.Err(), err)
+	} else if err != nil {
+		return nil, fmt.Errorf("accessing the remote image URL: %w", err)
+	}
+
+	layers, err := remoteImg.Layers()
+	if err != nil {
+		return nil, fmt.Errorf("obtaining the image layers: %w", err)
+	}
+
+	store, err := storage.GetStore(types.StoreOptions{
+		TransientStore: true,
+		RunRoot:        filepath.Join(os.TempDir(), "run"),
+		GraphRoot:      filepath.Join(os.TempDir(), "graph"),
+	})
+	if err != nil {
+		return nil, fmt.Errorf("opening storage: %w", err)
+	}
+
+	var parentIDs []string
+	for idx, layer := range layers {
+		rc, err := layer.Compressed() // Uncompressed?
```

Comment:
Interestingly, the `stream` package does not even implement `Uncompressed`: https://github.com/google/go-containerregistry/blob/59a4b85930392a30c39462519adc8a2026d47181/pkg/v1/stream/layer.go#L128

Now I do wonder what happens if a layer is not compressed to begin with, which is supported by the spec but uncommon, afaik.